### DISCUSSION TITLE
[General] Can not deploy a contract from another contract

### DISCUSSION THREAD

**Discussion by waynehoover** [zkSync-Community-Hub/zksync-developers#97] [Deployments] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I have solidity code that looks like this: `MyContract a = new MyContract{salt: ...}();` but I'm getting this error `execution reverted: The code hash is not known"` on that line.

I can't seem to narrow down the error, as it works for small trivial contracts.

It could be because the contract size is too big, because I get a size error when compiling but not a deploy to zksync testnet works.

### What is the specific issue or error you're encountering?

I can't deploy a contract from another contract. Here is a failed TX: https://goerli.explorer.zksync.io/tx/0x9130f38e5e167ee728adabe631d15e38c55b6fdabf02a015f68448bccb5a9429

Here is the code:
https://github.com/rabbitholegg/quest-protocol/pull/152/files#diff-6abf91e578b7f5e32753b99842c66f93d4246dbb4a587719bf9ebc7f806f7fa6R13

### Can you share the error messages or logs you're receiving, if any?

```
Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"execution reverted: The code hash is not known","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0x92e38e60c9e4e8aC6AcB24dE02d4a48856D775d4","maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"maxFeePerGas":{"type":"BigNumber","hex":"0x77359400"},"to":"0x49f05f2646772b6D0E695550CA49D1e3fF8A17E7","value":{"type":"BigNumber","hex":"0x00"},"data":"0x70dfd40a000000000000000000000000651c7fbbac1fa3a8937377c4879ca99c11f1261c0000000000000000000000000000000000000000000000000000000064d72df00000000000000000000000000000000000000000000000000000000064d5c0500000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000002433623438383835332d643631362d346162632d626662312d306335323065303630633934000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","type":2,"accessList":null},"error":{"code":3,"data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001a54686520636f64652068617368206973206e6f74206b6e6f776e000000000000"}}, tx={"data":"0x70dfd40a000000000000000000000000651c7fbbac1fa3a8937377c4879ca99c11f1261c0000000000000000000000000000000000000000000000000000000064d72df00000000000000000000000000000000000000000000000000000000064d5c0500000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000002433623438383835332d643631362d346162632d626662312d306335323065303630633934000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","to":{},"from":"0x92e38e60c9e4e8aC6AcB24dE02d4a48856D775d4","value":{"type":"BigNumber","hex":"0x00"},"type":2,"maxFeePerGas":{"type":"BigNumber","hex":"0x77359400"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x59682f00"},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.7.0)
```

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?

https://github.com/rabbitholegg/quest-protocol/pull/152/files#diff-6abf91e578b7f5e32753b99842c66f93d4246dbb4a587719bf9ebc7f806f7fa6R13

### Have you tried to isolate the problem, and if so, what were the results?

Yes, it's isolated to deploying a contract from another contract, but only here.

### What steps have you already taken to try to resolve the issue?

At first I thought it was related to this bug in the rust-sdk: https://github.com/matter-labs/zksync-era/issues/27 but probably not.

**Comment by uF4No** [comment]
hey @waynehoover 

I've created a minimal example that deploys one contract from another [in this repo](https://github.com/uF4No/zksync-contract-from-contract). I'm able to deploy a contract from another and interact with them with no issues. Please check it out in case you're doing something different.

**Comment by waynehoover** [comment]
Thanks @uF4No. I was also able to create a simple contract that deploys another contract, but for some reason, it's not working with my repo, it could be that the contract size is too big.

**Reply by Romsters** [reply]
Hi @waynehoover, we'll try to replicate it with your contract and will get back to you.

**Reply by Romsters** [reply, marked as answer]
Hi @waynehoover. I downloaded your code and added a simple function to `QuestFactoryBase.sol`:
```
function create1155Quest2(
        string memory questId_
    ) external payable nonReentrant returns (address) {
        address payable newQuest = payable(deploy1155Quest(erc1155QuestAddress, questId_));
        return newQuest;
    }
```
This one simply creates a contract within a contract. I changed nothing else and deployed it to zkSync Era testnet.
Transaction with calling this function was successfull and created a contract: https://goerli.explorer.zksync.io/tx/0x42a5163eba67e5c89090411a278da422f59d2d18fd2319081c94c04dc9470f74#eventlog.

**Comment by biotremenag** [comment]
The error "execution reverted: The code hash is not known" typically indicates an issue with the bytecode of the contract being deployed. It's possible that the contract size is indeed too big, as you mentioned encountering a size error during compilation.

To address this issue, you can try optimizing the contract code to reduce its size, such as removing unnecessary functions or using more efficient algorithms. Additionally, ensure that all external dependencies are correctly linked and that the contract bytecode is generated properly during compilation.

If the problem persists, consider breaking down the contract into smaller, more manageable components or deploying it in stages to identify the specific portion causing the error.

Lastly, ensure that you're using the latest versions of the tools and libraries involved, as updates may contain bug fixes or optimizations that could resolve the issue.

**Comment by natyfikapsha** [comment]
It seems like the issue might be related to the contract size, as you've mentioned encountering a size error during compilation. This could potentially affect the deployment process, leading to the "execution reverted: The code hash is not known" error. Additionally, deploying a contract from another contract may have its own complexities, especially if the contract size is large. It's worth investigating further to see if optimizing the contract size or deployment process could help resolve the issue.

**Comment by gauvinjohnny** [comment]
It seems like the error you're encountering, "The code hash is not known," might be due to the contract size being too large. This issue could potentially arise when deploying a contract from another contract. Additionally, you mentioned receiving a gas estimation error, which could also be related to the contract size. One potential solution could be to optimize the contract code to reduce its size. You may also want to consider breaking down the contract into smaller, more manageable components to facilitate deployment.
